### PR TITLE
Updates CMakeLists.txt to build for MSVC

### PR DESCRIPTION
**Pull request by toshipiazza**
Fixes some issues with compiler flags not supported on MSVC (-Werror)

Also, Protobuf builds by default with [/MT](https://github.com/protocolbuffers/protobuf/blob/d9ccd0c0e6bbda9bf4476088eeb46b02d7dcd327/examples/CMakeLists.txt#L16), whereas LPM built by default
with /MD. Now, when `-DLIB_PROTO_MUTATOR_DOWNLOAD_PROTOBUF=On` or
`-DLIB_PROTO_MUTATOR_MSVC_STATIC_RUNTIME=On` are specified we build LPM with
/MT so LPM and Protobuf can properly link together.

### CODE REVIEW

**Comment by toshipiazza** [comment]
Hmm, I thought I had signed the CLA once upon a time, a few years back... will TAL.

File: src/libfuzzer/libfuzzer_mutator.cc [hunk 1]
```diff
@@ -22,8 +22,7 @@
 #include "port/protobuf.h"
 #include "src/mutator.h"
 
-extern "C" size_t LLVMFuzzerMutate(uint8_t*, size_t, size_t)
-    __attribute__((weak));
+extern "C" size_t LLVMFuzzerMutate(uint8_t*, size_t, size_t);
```
Comment (toshipiazza) [inline src/libfuzzer/libfuzzer_mutator.cc]: LibFuzzer on MSVC doesn't use `__attribute__((weak))` (see [here](https://github.com/llvm-mirror/compiler-rt/blob/d798ef3964f20a6b76e513637c965ddeb8f7b3a0/lib/fuzzer/FuzzerExtFunctionsWindows.cpp#L35)). Either way, I don't think we need to declare this as a weak symbol for clang builds to be happy.

Comment (vitalybuka) [reply src/libfuzzer/libfuzzer_mutator.cc]: As-is it's going to break use-case when we test corpus without mutation.
In this case we don't link libfuzzer so symbol is missing. Should be true for windows as well.
We can avoid that if we split lib into parts where one part does not nee libfuzzer and another needs. But it will require some efforts.

File: CMakeLists.txt [hunk 2]
```diff
@@ -46,6 +46,21 @@ set(CMAKE_CXX_STANDARD 11)
 
 include_directories(${PROJECT_SOURCE_DIR})
 
+if (MSVC)
+  option(LIB_PROTO_MUTATOR_MSVC_STATIC_RUNTIME "Link static runtime libraries" ON)
+  if (LIB_PROTO_MUTATOR_DOWNLOAD_PROTOBUF OR LIB_PROTO_MUTATOR_MSVC_STATIC_RUNTIME)
+    # This is the default for protobuf with MSVC
+    # http://www.cmake.org/Wiki/CMake_FAQ#How_can_I_build_my_MSVC_application_with_a_static_runtime.3F
+    foreach(flag_var
```
Comment (vitalybuka) [inline CMakeLists.txt]: cmake part LGTM

BTW. would be nice to setup windows build on travis

File: src/libfuzzer/libfuzzer_mutator.cc (hunk 1 above)
Comment (toshipiazza) [reply src/libfuzzer/libfuzzer_mutator.cc]: Can you elaborate? With this change, all tests seem to build and pass on linux (tests fail to compile on Windows for other reasons, such as cmake failing to find libgtest.a--it's gtest.lib on Windows).

Tests link only against protobuf-mutator, zlib and gtest, so they don't seem to link against libfuzzer.

Comment (vitalybuka) [reply src/libfuzzer/libfuzzer_mutator.cc]: Unfortunately this case is not tested here, but it should be easy to test.

Let consider non libprotobuf-mutator case. You have a fuzzer e.g.
```
// fuzz_target.cc
extern "C" int LLVMFuzzerTestOneInput(const uint8_t *Data, size_t Size) {
  DoSomethingInterestingWithMyAPI(Data, Size);
  return 0;  // Non-zero return values are reserved for future use.
}
```

Usually you link it with libFuzzer and libFuzzer calls it.
Then you found interesting input which crashed something. When bug is fixed you may want to have some regression test for that e.g.:

```
// regression_test.cc
int main() {
  LLVMFuzzerTestOneInput("bad data", size);
}
```
in this case you don't need to link it with libFuzzer

Now lets get back to libprotobuf-mutator.
```
DEFINE_PROTO_FUZZER(const libfuzzer_example::Msg& message) {
    DoSomethingInterestingWithMyMessageI(message);
}
```
DEFINE_PROTO_FUZZER creates:
1. LLVMFuzzerTestOneInput which just parses a message.
2. LLVMFuzzerCustomMutator/Crossover which mutate a message

Our regression tests are still very similar:
```
int main() {
  LLVMFuzzerTestOneInput("bad serialized message", size);
}
```
It calls LLVMFuzzerTestOneInput from DEFINE_PROTO_FUZZER, but there is unneeded LLVMFuzzerCustomMutator/Crossover which need LLVMFuzzerMutate.

This is how we run regression tests inside of Google, Chromium should be setup same way.
main() is more complicated, It loads inputs from file but idea is the same.

So how to avoid that without weak?
I guess we can just link dummy LLVMFuzzerMutate if we don't link libfuzzer.
Let me try this on our code, if it works we will merge the patch as-is and I cover this case with a test or/and documentation.

**Comment by kcc** [comment]
Is there any other way to implement a weak function on Windows? 
I really want to avoid having to define `LLVMFuzzerMutate` anywhere else

File: src/libfuzzer/libfuzzer_mutator.cc (hunk 1 above)
Comment (toshipiazza) [reply src/libfuzzer/libfuzzer_mutator.cc]: If this doesn't work, I can put back the `__attribute__((weak))` and put it behind an ifdef. I can also test this exact scenario out on Windows this weekend; I suspect something like this might work:

```c
#ifdef LIBFUZZER_MSVC
  extern "C" size_t LLVMFuzzerMutateDef(uint8_t*, size_t, size_t);
  __pragma(comment(linker, "/alternatename:LLVMFuzzerMutate=LLVMFuzzerMutateDef"));
#else
  extern "C" size_t LLVMFuzzerMutate(uint8_t*, size_t, size_t)
    __attribute__((weak));
#endif
```

**Comment by toshipiazza** [comment]
@kcc see my comment [above](https://github.com/google/libprotobuf-mutator/pull/125#discussion_r289179887)

**Comment by kcc** [comment]
Is something like this possible? 

```cpp
#ifdef MSVC
# define ATTRIBUTE_WEAK SomeMagic
#else 
# define ATTRIBUTE_WEAK __attribute__((weak))
#endif 
```

**Comment by toshipiazza** [comment]
@kcc I'm not aware of anything besides /alternatename that mimics `__attribute__((weak))`, even after doing some research.

File: src/libfuzzer/libfuzzer_mutator.cc (hunk 1 above)
Comment (vitalybuka) [reply src/libfuzzer/libfuzzer_mutator.cc]:
```
#ifdef MSVC
# define ATTRIBUTE_WEAK ;__pragma(comment(linker, "/alternatename:LLVMFuzzerMutate=LLVMFuzzerMutateDef"));
#else 
# define ATTRIBUTE_WEAK __attribute__((weak))
#endif 
```

Comment (vitalybuka) [reply src/libfuzzer/libfuzzer_mutator.cc]: Maybe something like this?
https://github.com/llvm/llvm-project/blob/885b790f89b6068ec4caad8eaa51aa8098327059/compiler-rt/lib/sanitizer_common/sanitizer_internal_defs.h#L56

Comment (toshipiazza) [reply src/libfuzzer/libfuzzer_mutator.cc]: Hmm, that looks to eventually do the same thing:

```
#define WIN_WEAK_ALIAS(Name, Default)                                          \
  __pragma(comment(linker, "/alternatename:" WIN_SYM_PREFIX STRINGIFY(Name) "="\
                                             WIN_SYM_PREFIX STRINGIFY(Default)))
```

Comment (vitalybuka) [reply src/libfuzzer/libfuzzer_mutator.cc]: Yes, my point is that we can't just define ATTRIBUTE_WEAK and we need macro with name argument

Comment (toshipiazza) [reply src/libfuzzer/libfuzzer_mutator.cc]: Would you like me to update this PR with a change for this?

File: cmake/external/googletest.cmake [hunk 3]
```diff
@@ -23,11 +23,20 @@ set(GTEST_MAIN_LIBRARIES gtest_main)
 set(GTEST_BOTH_LIBRARIES ${GTEST_LIBRARIES} ${GTEST_MAIN_LIBRARIES})
 
 foreach(lib IN LISTS GTEST_BOTH_LIBRARIES)
-  list(APPEND GTEST_BUILD_BYPRODUCTS ${GTEST_INSTALL_DIR}/lib/lib${lib}.a)
+  if (MSVC)
+    if (CMAKE_BUILD_TYPE MATCHES Debug)
```
Comment (toshipiazza) [inline cmake/external/googletest.cmake]: I'm not a fan of this. This will work for the common case:

```
cmake -DCMAKE_BUILD_TYPE=Debug ..
cmake --build .
```

However, MSVC targets also allow you to specify the target at build time:

```
cmake ..
cmake --build . --config Debug # this will fail looking for gtest.lib when only gtestd.lib exists
```

Should I open another issue? IMO, a more permanent solution would be https://crascit.com/2015/07/25/cmake-gtest/. This is also called out on the gtest README proper.

File: src/libfuzzer/libfuzzer_mutator.cc (hunk 1 above)
Comment (vitalybuka) [reply src/libfuzzer/libfuzzer_mutator.cc]: yes please

File: cmake/external/googletest.cmake (hunk 3 above)
Comment (vitalybuka) [reply cmake/external/googletest.cmake]: which part of https://crascit.com/2015/07/25/cmake-gtest/ are you referring to?

Comment (toshipiazza) [reply cmake/external/googletest.cmake]: The post describes creating a very simple intermediate CMakeLists.txt.in which calls ExternalProject_Add on gtest; then the main CMakeLists.txt runs:

```cmake
# Download and unpack googletest at configure time
configure_file(CMakeLists.txt.in googletest-download/CMakeLists.txt)
execute_process(COMMAND "${CMAKE_COMMAND}" -G "${CMAKE_GENERATOR}" .
    WORKING_DIRECTORY "${CMAKE_BINARY_DIR}/googletest-download" )
execute_process(COMMAND "${CMAKE_COMMAND}" --build .
    WORKING_DIRECTORY "${CMAKE_BINARY_DIR}/googletest-download" )

...

# Add googletest directly to our build. This adds
# the following targets: gtest, gtest_main, gmock
# and gmock_main
add_subdirectory("${CMAKE_BINARY_DIR}/googletest-src"
                 "${CMAKE_BINARY_DIR}/googletest-build")
```

According to the post, the gtest_main and gtest targets are now directly available, so we don't have to reference build artifacts by name out of the gtest build dir, which is architecture+toolchain dependant. This is a little hacky, but it's also documented in the [googletest](https://github.com/google/googletest/tree/master/googletest#incorporating-into-an-existing-cmake-project) repo as well.

Comment (vitalybuka) [reply cmake/external/googletest.cmake]: lgtm

File: src/libfuzzer/libfuzzer_mutator.cc (hunk 1 above)
Comment (toshipiazza) [reply src/libfuzzer/libfuzzer_mutator.cc]: Is there anything more that should be done for this?

File: src/utf8_fix_test.cc [hunk 4]
```diff
@@ -45,7 +45,7 @@ INSTANTIATE_TEST_SUITE_P(Size, FixUtf8StringTest, ::testing::Range(0, 10));
 
 TEST_P(FixUtf8StringTest, FixUtf8String) {
   RandomEngine random(GetParam());
-  std::uniform_int_distribution<uint8_t> random8(0, 0xFF);
+  std::uniform_int_distribution<> random8(0, 0xFF);
```
Comment (vitalybuka) [inline src/utf8_fix_test.cc]: Why this change?

Comment (vitalybuka) [reply src/utf8_fix_test.cc]: if not utf8 then std::string::value_type?

File: src/libfuzzer/libfuzzer_mutator.cc [hunk 5]
```diff
@@ -22,8 +22,33 @@
 #include "port/protobuf.h"
 #include "src/mutator.h"
 
+#ifdef _MSC_VER
```
Comment (vitalybuka) [inline src/libfuzzer/libfuzzer_mutator.cc]: with this ifdef WIN_WEAK_ALIAS make not much sense, you can keep just __pragma

I would prefere have common code with equivalent of SANITIZER_INTERFACE_WEAK_DEF
and it will be expanded into platform specific stuff

File: src/utf8_fix_test.cc (hunk 4 above)
Comment (toshipiazza) [reply src/utf8_fix_test.cc]: According to https://en.cppreference.com/w/cpp/numeric/random/uniform_int_distribution:

>  The result type generated by the generator. The effect is undefined if this is not one of short, int, long, long long, unsigned short, unsigned int, unsigned long, or unsigned long long

MSVC also fails to compile this with uint8_t as the template argument.

Should I cast to uint8_t wherever the random8 generator is used here and elsewhere?

Comment (vitalybuka) [reply src/utf8_fix_test.cc]: thanks for explanantion
it's fine with implicit cast

File: src/libfuzzer/libfuzzer_mutator.cc (hunk 5 above)
Comment (toshipiazza) [reply src/libfuzzer/libfuzzer_mutator.cc]: Done.
